### PR TITLE
Add session pool.

### PR DESCRIPTION
**Pull request by Shylock-Hg**

### CODE REVIEW

File: include/nebula/client/SessionPool.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      if (resp.errorCode == ErrorCode::SUCCEEDED) {
+        idleSessions_.emplace_back(std::move(session));
+      }
+    }
+  }
+
+  // Session pool use fixed space, don't switch space when execute
+  ExecutionResponse execute(const std::string &stmt);
+
+  ExecutionResponse executeWithParameter(const std::string &stmt,
```
Comment (Aiee) [inline include/nebula/client/SessionPool.h:53]: What if the user executes a `use xxx` query?

Comment (Shylock-Hg) [reply include/nebula/client/SessionPool.h:53]: ok

File: include/nebula/client/SessionPool.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+#include <mutex>
+
+#include "nebula/client/Config.h"
+#include "nebula/client/ConnectionPool.h"
+#include "nebula/client/Session.h"
+
+namespace nebula {
+
+struct SessionPoolConfig {
+  std::string username_;
```
Comment (Aiee) [inline include/nebula/client/SessionPool.h:20]: Add checks to validate the config, such as the username/password should not be empty

Comment (Shylock-Hg) [reply include/nebula/client/SessionPool.h:20]: ok

File: include/nebula/client/SessionPool.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  std::string spaceName_;
+  // Socket timeout and Socket connection timeout, unit: seconds
+  std::uint32_t timeout_{0};
+  // The idleTime of the connection, unit: seconds
+  // If connection's idle time is longer than idleTime, it will be delete
+  // 0 value means the connection will not expire
+  std::uint32_t idleTime_{0};
+  std::uint32_t maxSize_{10};  // max size of the session pool. should be adjusted according to the
+                               // max threads will be using.
+  std::uint32_t minSize_{0};   // min size of  the session pool
```
Comment (Aiee) [inline include/nebula/client/SessionPool.h]: set default value to 1

File: src/client/SessionPool.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  pool_->init(config_.addrs_, conf);
+  if (config_.spaceName_.empty()) {
+    return false;
+  }
+  if (config_.username_.empty() || config_.password_.empty()) {
+    return false;
+  }
+  std::string useSpace = "USE " + config_.spaceName_;
+  for (std::size_t i = 0; i < config_.maxSize_; ++i) {
+    // use space
```
Comment (Aiee) [inline src/client/SessionPool.cpp:29]: i < config_.minSize_

File: src/client/tests/SessionPoolTest.cpp [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+  nebula::SessionPoolConfig config;
+  config.username_ = "root";
+  config.password_ = "nebula";
+  config.addrs_ = {kServerHost ":9669"};
+  config.spaceName_ = "session_pool_test";
+  config.maxSize_ = 4;
+  nebula::SessionPool pool(config);
+  pool.init();
+  // switch space and auto switch back
+  {
```
Comment (Aiee) [inline src/client/tests/SessionPoolTest.cpp:166]: If you do not run  `USE session_pool_test2` and  `YIELD 1` in the same query, they might run via different sessions.

Simply do `execute("USE session_pool_test2; YIELD 1")` and check the space in the response. We do not need multithreaded here

File: include/nebula/client/SessionPool.h (hunk 3 above)
Comment (Shylock-Hg) [reply include/nebula/client/SessionPool.h]: ok

File: src/client/tests/SessionPoolTest.cpp (hunk 5 above)
Comment (Shylock-Hg) [reply src/client/tests/SessionPoolTest.cpp:166]: Good catch.

File: src/client/SessionPool.cpp (hunk 4 above)
Comment (Shylock-Hg) [reply src/client/SessionPool.cpp:29]: Emm. It's a legacy of nebula-cpp. I will fix in later.

**Review by jievince — APPROVED** [review]

**Pull request merged**